include/rsp_queue.inc

```diff
@@ -0,0 +1,566 @@
+#    back over to the queue engine. The return register will always
+#    point to the main loop entry point when calling your commands,
+#    so you can simply do this: `jr ra`. Instead, you can also 
+#    explicitly jump there: `j RSPQ_Loop`.
+# 4. Somewhere in your data segment, define the saved overlay
+#    state using RSPQ_BeginSavedState and RSPQ_EndSavedState.
+#    Make sure to put at least one data directive that emits
+#    a non-zero amount of bytes between them. If your overlay
+#    doesn't need any data to be persisted, then use
+#    RSPQ_EmptySavedState instead.
+# 
+# Read below for more details on how to use the macros mentioned above.
+#
+########################################################
+
+########################################################
+# RSPQ_BeginOverlayHeader
+# 
+# Begins the definition of an overlay header. Must be
+# followed by at least one call to RSPQ_DefineCommand
+# and closed with RSPQ_EndOverlayHeader. Only calls
+# to RSPQ_DefineCommand are allowed inside the header definition.
+########################################################
+.macro RSPQ_BeginOverlayHeader
+_RSPQ_OVERLAY_HEADER:
+    # state start
+    .short _RSPQ_SAVED_STATE_START - _data_start
+    # state size
+    .short _RSPQ_SAVED_STATE_END - _RSPQ_SAVED_STATE_START - 1
+    # command base (filled in by C code)
+    .short 0
+    # unused padding
+    .short 0
+    
+	.align 1
+_RSPQ_OVERLAY_COMMAND_TABLE:
+.endm
+
+########################################################
+# RSPQ_BeginSavedState
+# 
+# Begins the definition of the saved overlay state. Must be
+# closed by RSPQ_EndSavedState. The region of memory between 
+# RSPQ_BeginSavedState and RSPQ_EndSavedState will be preserved 
+# by the queue engine across all calls to an overlay, even if 
+# it gets unloaded from memory.
+#
+# Every overlay must define exactly one saved state and it
+# must contain at least one data directive or, in other words,
+# its size must not be zero.
+########################################################
+.macro RSPQ_BeginSavedState
+    .align 3
+_RSPQ_SAVED_STATE_START:
+.endm
+
+########################################################
+# RSPQ_EndSavedState
+# 
+# Ends the definition of the saved overlay state.
+########################################################
+.macro RSPQ_EndSavedState
+    .align 3
+_RSPQ_SAVED_STATE_END:
+.endm
+
+########################################################
+# RSPQ_EmptySavedState
+# 
+# Defines an empty saved overlay state. Use this instead
+# of RSPQ_BeginSavedState and RSPQ_EndSavedState if your
+# overlay does not need any persistent state.
+#
+# It is only possible to use either this or 
+# RSPQ_BeginSavedState/RSPQ_EndSavedState, but not both.
+# Do not use RSPQ_BeginSavedState and RSPQ_EndSavedState
+# without any data directives between them.
+########################################################
+.macro RSPQ_EmptySavedState
+    RSPQ_BeginSavedState
+    .quad 0
```

Comment:
If the size could be zero, we'd need an additional branch for saving the state to RDRAM (rsp_queue.inc, lines 326-328). Also because the state size is stored as `size - 1` (so it can be used for DMA straight away), a size of zero would cause an overflow, which in turn causes undefined behavior because the RSP would attempt to DMA 0xFFFF + 1 = 4k of DMEM (so all of it I guess).

In short: This is an artificial restriction so we can save yet another instruction :)